site/examples/code-highlighting.tsx

```diff
@@ -1,237 +1,589 @@
-        return ranges
+import { ParagraphElement, CodeLineElement } from './custom-types'
+
+const ParagraphType = 'paragraph'
+const CodeLineType = 'code-line'
+
+const renderElement = (props: RenderElementProps) => {
+  const { attributes, children, element } = props
+  const editor = useSlateStatic()
+  const Tag = editor.isInline(element) ? 'span' : 'div'
+  const attrs = {
+    'data-slate-element-type': element.type,
+  }
+
+  if (element.type === CodeLineType) {
+    const intervalsStarts =
+      editor.intervalsStarts.get(CodeLineType) || new Set()
+    const path = ReactEditor.findPath(editor, element)
+    const selectVisible = intervalsStarts.has(path[0])
+
+    const setLanguage = (language: string) => {
+      const intervals = editor.intervals.get(CodeLineType) || []
+      const interval = intervals.find(
+        ([start, end]) => start <= path[0] && end >= path[0]
+      )
+      if (interval) {
+        Transforms.setNodes(
+          editor,
+          { language },
+          {
+            at: {
+              anchor: Editor.start(editor, [interval[0]]),
+              focus: Editor.end(editor, [interval[1]]),
+            },
+          }
+        )
       }
-      const tokens = Prism.tokenize(node.text, Prism.languages[language])
-      let start = 0
+    }
+
+    return (
+      <Tag
+        {...attributes}
+        {...attrs}
+        style={{ position: 'relative' }}
+        className={
+          element.type === CodeLineType &&
+          css(`
+        font-family: monospace;
+        font-size: 16px;
+        line-height: 20px;
+        margin-top: 0;
+        background: rgba(0, 20, 60, .03);
+        padding: 5px 13px;
+      `)
+        }
+        spellCheck={element.type !== CodeLineType}
+      >
+        {selectVisible && (
+          <LanguageSelect
+            value={element.language}
+            onChange={e => setLanguage(e.target.value)}
+          />
+        )}
+        {children}
+      </Tag>
+    )
+  }
+
+  return (
+    <Tag {...attributes} {...attrs} style={{ position: 'relative' }}>
+      {children}
+    </Tag>
+  )
+}
+
+const renderLeaf = (props: RenderLeafProps) => {
+  const { attributes, children, leaf } = props
+  const { text, ...rest } = leaf
+
+  return (
+    <span {...attributes} className={Object.keys(rest).join(' ')}>
+      {children}
+    </span>
+  )
+}
+
+const useDecorate = (editor: Editor) => {
+  return useCallback(([_, path]) => {
+    if (path.length !== 1) {
+      // skip not top level elements
+      return []
+    }
+
+    // find ranges for element by id
+    const element = editor.children[path[0]]
+    const id = ReactEditor.findKey(editor, element).id
+    const ranges = editor.ranges.get(id)
+
+    return ranges ? ranges.slice() : []
+  }, [])
+}
+
+// set editor.intervals with intervals of a continuous stream of identical elements
+// set editor.intervalsStarts with set of these intervals start index (for optimization)
+const ExtractIntervals = () => {
```

Comment:
could add this as a hook instead perhaps?

Reply:
So on every editor/selection change this loops over all (top-level only?) editor children. Could you explain what an interval is?

Reply:
Hey @bryanph. Yes, it's just the contiguous elements intervals calculated on each editor/selection change And it loops only over all top-level editor children. It may have the following data:

```json
{ "paragraph": [[0, 0], [12,14]], "code-line": [[1, 11], [15, 25]] }
```

So the idea here is to extract additional information based on the current editor's children state. It should be always synced with editor children. That's why these "extractors" were placed inside Slate, but before Editable. It means that all components inside Editable can rely on `editor.intervals` as on `editor.children`. I think it actually can be reorganised with hooks in some way. I just wanted to keep it simple.

Reply:
Actually, it is could be optimized by moving these calculations into

```javascript
useMemo(() => {...; editor.intervals = intervals; }, [editor.children])
```

hook to prevent calling these calculations when only selection is changed